### PR TITLE
hal-hardware-analyzer: 3.3.0 -> 4.2.0

### PR DESCRIPTION
**Pull request by risicle**
## Description of changes
This is largely the work of @doronbehar who wrote most of the de-vendoring patch used here (attempted upstreaming @ https://github.com/emsec/hal/pull/529)

I still can't get this working on x86_64 darwin - it uses `open_memstream` and I can't get `memstreamHook` from #135525 to fully solve the issues.

But seeing as hal is currently totally broken on darwin, this is at least an improvement. Addresses #240729

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix [hunk 1]
```diff
@@ -1,90 +1,100 @@
...
+        url = "https://github.com/igraph/igraph/commit/5ad464be5ae2f6ebb69c97cb0140c800cc8d97d6.patch";
+        hash = "sha256-adU5SctH+H54UaAmr5BZInytD3wjUzLtQbCwngAWs4o=";
+      })
+    ];
+    postPatch = prev.postPatch + lib.optionalString stdenv.isAarch64 ''
       # https://github.com/igraph/igraph/issues/1694
       substituteInPlace tests/CMakeLists.txt \
         --replace "igraph_scg_grouping3" "" \
         --replace "igraph_scg_semiprojectors2" ""
     '';
```
Comment (doronbehar) [inline pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:45]: Is this `postPatch` relevant also for igraph 0.10.6?

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix [hunk 2]
```diff
@@ -1,90 +1,100 @@
...
-, graphviz
-, llvmPackages
 , z3
-, fmt_8
-, suitesparse
 }:
 
 let
-  igraph' = igraph.overrideAttrs (old: rec {
+  igraph' = igraph.overrideAttrs (final: prev: {
```
Comment (doronbehar) [inline pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:24]: It'd be nice if upstream was aware that there is an igraph version which they don't support.

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix [hunk 3]
```diff
@@ -1,90 +1,100 @@
...
+    postPatch = prev.postPatch + lib.optionalString stdenv.isAarch64 ''
       # https://github.com/igraph/igraph/issues/1694
       substituteInPlace tests/CMakeLists.txt \
         --replace "igraph_scg_grouping3" "" \
         --replace "igraph_scg_semiprojectors2" ""
     '';
-    buildInputs = old.buildInputs ++ [ suitesparse ];
-    cmakeFlags = old.cmakeFlags ++ [ "-DIGRAPH_USE_INTERNAL_CXSPARSE=OFF" ];
+    buildInputs = prev.buildInputs ++ [ suitesparse ];
+    cmakeFlags = prev.cmakeFlags ++ [ "-DIGRAPH_USE_INTERNAL_CXSPARSE=OFF" ];
```
Comment (doronbehar) [inline pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:47]: These features are specific to HAL? I also wonder whether they should be part of the main expression.

**Review by doronbehar — COMMENTED** [review]
Looks pretty good overall!

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix (hunk 1 above)
Comment (szhorvat) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:45]: No, it is not. 

But I thought `hal` was not prepared for igraph 0.10 yet, see https://github.com/emsec/hal/pull/487

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix (hunk 2 above)
Comment (szhorvat) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:24]: They are aware, see https://github.com/emsec/hal/pull/487

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix (hunk 3 above)
Comment (risicle) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:47]: Yeah I think these were just "brought back" from the 0.9.x package to make this overridden version as similar as possible to the old 0.9.x package.

Comment (doronbehar) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:47]: > Yeah I think these were just "brought back" from the 0.9.x package to make this overridden version as similar as possible to the old 0.9.x package.

It'd be nice to write that in a comment.

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix (hunk 2 above)
Comment (doronbehar) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:24]: > They are aware, see [emsec/hal#487](https://github.com/emsec/hal/pull/487)

Thanks for the link :) @risicle could you mention this in a comment as well?

File: pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix (hunk 1 above)
Comment (doronbehar) [reply pkgs/applications/science/electronics/hal-hardware-analyzer/default.nix:45]: Same here @risicle , it'd be nice to be track down those upstream issues via a comment link in our expression.

**Comment by risicle** [comment]
Looking closer at the built result of this, the `graph_algorithms.so` plugin is broken because it doesn't link with igraph. Will have to dig deeper.

**Comment by risicle** [comment]
If @tobim wants to pull this in to #250435 instead that would be fine too.

**Comment by tobim** [comment]
> If @tobim wants to pull this in to #250435 instead that would be fine too.

Nono, I just cherry-picked an earlier version of this PR to fix eval. Please continue.

**Review by doronbehar — APPROVED** [review]
Changes look good, haven't tested. I trust OP have tested them.

**Pull request merged**